### PR TITLE
Assign unknown client IPs a random geo ip in the contiguous US & cache the result

### LINKED ISSUE #1587: Generate & cache geolocation for uknown client IPs
Right now, the Director generates a completely random list of caches when redirecting a client whose IP has no GeoIP resolution. As our understanding of how cache selection should work evolves, we've identified two primary issues with this:

1. The same GeoIP-less client making repeated requests to the Director gets sent to different caches every time, severely affecting performance (more or less negating the benefits of caching in the first place). Instead of taking this approach, the Director should pick a random GeoIP location (assumed to be in the US) and cache that result for ~20 minutes. This allows repeated requests to get the same list of caches. Note that this moves a lot of the randomness out of the cache selection, and into the GeoIP assignment the Director makes for the client.
2. GeoIP-less clients may sometimes be redirected to caches outside the US, but we'd rather everyone get sent to something in the US. By fixing 1, we also fix 2.

### PR DESCRIPTION
**Pull request by jhiemstrawisc**
This PR aims to improve the Director's cache sorting algorithm, specifically in the case of clients whose GeoIP resolution can't be determined.

With these changes, client IPs that don't have an entry in MaxMind entry (or that can't be parsed as an IP in the first case) are assigned a random location in the contiguous United States. Doing this makes the Director behave as though it knows the client's true location by delivering caches sorted based on the new location. This differs from the previous algorithm, where those IP addresses resulted in completely random cache sorting.

When a client IP is assigned a random location, that location is cached for 20 minutes, unless refreshed by further client interactions. This is done so that repetitive client requests get a consistent cache ordering (barring any other load information or the sort algorithm's stochastic jiggle), which should further improve cache hits.

Closes #1587

### CODE REVIEW

File: director/sort_test.go [hunk 1]
```diff
@@ -450,11 +450,20 @@ func TestGetClientLatLong(t *testing.T) {
...
 		coord := getClientLatLong(clientIp)
 
 		assert.True(t, coord.Lat <= usLatMax && coord.Lat >= usLatMin)
 		assert.True(t, coord.Long <= usLongMax && coord.Long >= usLongMin)
 		assert.Contains(t, logOutput.String(), "Unable to sort servers based on client-server distance. Invalid client IP address")
+		assert.NotContains(t, logOutput.String(), "Retrieving pre-assigned lat/long")
+
+		// Get it again to make sure it's coming from the cache
+		coord = getClientLatLong(clientIp)
+		assert.True(t, coord.Lat <= usLatMax && coord.Lat >= usLatMin)
```
Comment (turetske) [inline director/sort_test.go:468]: Rather than asserting that it's within the same  US bound, assert that the two coordinates are the same.

**Review by turetske — CHANGES_REQUESTED** [review]
One minor change to a test requested.

File: director/sort_test.go (hunk 1 above)
Comment (jhiemstrawisc) [reply director/sort_test.go]: Good call. Done :)

**Review by turetske — APPROVED** [review]
LGTM!

**Pull request merged**